src/Val.h

```diff
@@ -327,13 +333,13 @@ class ValManager
 	inline const StringValPtr& EmptyString() const { return empty_string; }
 
 	// Port number given in host order.
-	const PortValPtr& Port(uint32_t port_num, TransportProto port_type) const;
+	const PortValPtr& Port(uint32_t port_num, TransportProto port_type);
 
 	// Host-order port number already masked with port space protocol mask.
-	const PortValPtr& Port(uint32_t port_num) const;
+	const PortValPtr& Port(uint32_t port_num);
 
 private:
-	std::array<std::array<PortValPtr, 65536>, NUM_PORT_SPACES> ports;
+	std::map<std::pair<uint32_t, TransportProto>, PortValPtr> ports;
```

Comment:
Should we consider switch this to 1) `std::unordered_map` and 2) possibly to just `<uint32_t, PortValPtr>` where the `uint32_t` value is already the port space masked value?

A naive benchmark indicates `unordered_map` being considerably faster once there are enough elements in the map, albeit slower for few elements.
```
-------------------------------------------------------------------------------
Benchmark                                     Time             CPU   Iterations
-------------------------------------------------------------------------------
std_map_std_pair/32                        7.90 ns         7.90 ns     91855437
std_map_std_pair/64                        8.13 ns         8.13 ns     85607659
std_map_std_pair/512                       31.0 ns         31.0 ns     22837893
std_map_std_pair/1024                      48.0 ns         48.0 ns     14880275
std_unordered_map_std_pair/32              14.4 ns         14.4 ns     50066974
std_unordered_map_std_pair/64              14.2 ns         14.2 ns     49192062
std_unordered_map_std_pair/512             14.7 ns         14.7 ns     47303798
std_unordered_map_std_pair/1024            15.6 ns         15.6 ns     44525559
std_unordered_map_just_uint32_t/32         11.1 ns         11.1 ns     62767385
std_unordered_map_just_uint32_t/64         9.26 ns         9.26 ns     75598475
std_unordered_map_just_uint32_t/512        12.0 ns         12.0 ns     58354134
std_unordered_map_just_uint32_t/1024       12.7 ns         12.7 ns     54651908
```

```
#include <benchmark/benchmark.h>

#include <cstdint>
#include <map>
#include <string>
#include <unordered_map>

struct pair_hash
{
    template <class T1, class T2>
    std::size_t operator() (const std::pair<T1, T2> &pair) const {
        return std::hash<T1>()(pair.first) ^ std::hash<T2>()(pair.second);
    }
};

#define TCP_PORT_MASK 0x10000
#define UDP_PORT_MASK 0x20000
#define ICMP_PORT_MASK 0x30000

enum TransportProto
        {
        TRANSPORT_UNKNOWN,
        TRANSPORT_TCP,
        TRANSPORT_UDP,
        TRANSPORT_ICMP,
        };


static void std_map_std_pair(benchmark::State& state) {
  std::map<std::pair<uint32_t, TransportProto>, uint32_t> map;
  const int map_sz = state.range(0);

  for (int i = 0; i < map_sz; i++) {
    const auto k = std::pair<uint32_t, TransportProto>(i, TransportProto(i % 3 + 1));
    map[k] = i;
  }

  int i = 0;
  for (auto _ : state) {
    i = (i + 13) % map_sz;
    const auto k = std::pair<uint32_t, TransportProto>(i, TransportProto(i % 3 + 1));
    benchmark::DoNotOptimize(map[k]);
    benchmark::ClobberMemory();
  }
}

static void std_unordered_map_std_pair(benchmark::State& state) {
  std::unordered_map<std::pair<uint32_t, TransportProto>, uint32_t, pair_hash> map;
  const int map_sz = state.range(0);

  for (int i = 0; i < map_sz; i++) {
    const auto k = std::pair<uint32_t, TransportProto>(i, TransportProto(i % 3 + 1));
    map[k] = i;
  }

  int i = 0;
  for (auto _ : state) {
    i = (i + 13) % map_sz;
    const auto k = std::pair<uint32_t, TransportProto>(i, TransportProto(i % 3 + 1));
    benchmark::DoNotOptimize(map[k]);
    benchmark::ClobberMemory();
  }
}

static void std_unordered_map_just_uint32_t(benchmark::State& state) {
  std::unordered_map<uint32_t, int32_t> map;
  const int map_sz = state.range(0);

  for (int i = 0; i < map_sz; i++) {
    const uint32_t k = i | (i % 2 == 0 ? TCP_PORT_MASK : UDP_PORT_MASK);
    map[k] = i;
  }

  int i = 0;
  for (auto _ : state) {
    i = (i + 13) % map_sz;
    const uint32_t k = i | (i % 2 == 0 ? TCP_PORT_MASK : UDP_PORT_MASK);
    benchmark::DoNotOptimize(map[k]);
    benchmark::ClobberMemory();
  }
}

BENCHMARK(std_map_std_pair)->Range(32, 8<<7);
BENCHMARK(std_unordered_map_std_pair)->Range(32, 8<<7);
BENCHMARK(std_unordered_map_just_uint32_t)->Range(32, 8<<7);

BENCHMARK_MAIN();
```

Reply:
The original std::array version appears to run twice as fast as the std::unordered_map/uint32_t version still. It is a microbenchmark, but PortVal lookups aren't exactly uncommon.

Reply:
> The original std::array version appears to run twice as fast as the std::unordered_map/uint32_t version still.

That's a pretty stark difference. I'm not sure that I feel comfortable losing that performance difference for the majority of users.

Reply:
Let's go ahead and change this over to the `std_unordered_map_just_uint32_t` recommended version here, and we can write an issue to revisit the performance difference between that and `std::array` and how to handle it.

Reply:
Done!